### ISSUE TITLE
Correlated data in network training

### ISSUE THREAD

**Issue by jkiliani** [leela-zero/leela-zero#1197]
I'm going to post a number of observations about training here that led me to a hypothesis about network training that affects Leela Zero, Leela Chess and Minigo.

For Leela Zero, during regular reinforcement learning, we never had overfitting problems (after the shuffle buffer fix at least). With the last bootstrap, we did however, until the window size was increased (pioneered by @bjiyxo and adopted by @gcp)

For Leela chess, reinforcement learning with window sizes of 250-300k typically ended up with a considerable train-test variance in MSE, indicating that the value head was generally overfit to an extent, although not enough that it seriously degraded performance. Recently, several bootstrap nets were trained, first a 128x10 net on 600k games, then both a 64x6 net and another 128x10 net on 1.5 million games each. The networks with the much larger window has a lot lower train-test variances, especially the 64x6 net.

For Minigo, performance appears stalled, which is strange considering the 256x20 architecture.

Except for Leela Zero regular reinforcement learning (which uses gating), the commonality I see here is that bigger windows tend to work better than smaller ones.

I think the reason for this is that while we looked at the data correlation within mini-batches a while ago (the shuffle buffer fix), we didn't really investigate the effect of data correlation within the training window. The number of really independent data points going into the value head training is actually only as large as the number of games, since we train each position on the game outcome. I think that is the main reason that we had failed bootstraps, and that always-promote doesn't work well on a static window size. There must be a certain number of positions per game that you can safely train on during one training run without overfitting the value head, and it appears the successful bootstraps stayed below that threshold while the failed bootstraps didn't.

The current gating procedure in Leela Zero protects the network from the effects of value head overfitting simply by making sure that overfit nets won't pass gating, and the reset to best network removes any overfitting for the next training run. However, with always-promote we would give up that safeguard, and likely have to use window sizes comparable to the last bootstrap (1.5 million?) to not end up with an overfit value head. Furthermore, with every architecture increase, the training window would have to be increased as well to avoid value head overfitting.

While the current training scheme clearly works, I'd like to propose a new training schedule to @gcp that would refine the current method:

- Keep resetting the net to current best network after each run
- Make gating even stricter, so it's very certain that a passed network really is significantly better, maybe 60% at 400 games?
- Instead of using fixed training window size, make the window size adaptive: After each promotion, operate on the assumption that the net was trained on the data within the current window as much as is actually helpful, and reset the window to zero (i.e. remove all games that went into training the successful net, and train only on those produced afterwards)
- Then train with a step count schedule where both the total training steps and the steps at which we test are proportional to window size. I.e. after a reset, do very short training runs and test often, while if there are many games since the last promotion, train for more steps, and test less often. The reason behind this is that we should try to keep the number of positions trained on each game roughly similar throughout the pipeline.
- It may save a lot of resources to not actually test all that many points, but use the training accuracy and test MSE scores as indicators which step sizes are worth testing, and when the run has definitely started to overfit and can be reset. It may NOT be optimal to choose the test set from the same window as the training set by the way. A test set picked from a much larger window would likely work better, if test accuracy and MSE are used to determine what to test and when to reset.

TLDR: We're currently not taking the correlation of data from the same games within the training window into account, although that hurt mostly when doing bootstraps, since gating provides a working safeguard. Regardless, the training procedure with gating has potential for refinement.

**Comment by Marcin1960** [comment]
@jkiliani "Instead of using fixed training window size, make the window size adaptive: After each promotion, operate on the assumption that the net was trained on the data within the current window as much as is actually helpful, and reset the window to zero (i.e. remove all games that went into training the successful net, and train only on those produced afterwards)"

Very interesting!

**Comment by herazul** [comment]
It could be interesting. However, i don't see why it would be interesting to make gating higher ? we saw that 55% is enough to not get any problems.

**Comment by jkiliani** [comment]
@herazul I agree that 55% is not likely to run into actual problem, but the main reason I proposed stricter gating is the window reset: If we have a net passing due to luck instead of skill, the games up to that point wouldn't be used anymore undeservedly.

But agreed that maybe this part could still use tweaking.

**Comment by isty2e** [comment]
The validation set, which is currently sampled from the training window, also should be large enough. Otherwise validation will be quite noisy and unreliable. And there are some questions to be addressed for a fruitful discussion:

1. Do we actually experience overfitting with the current learning rate and steps? We have to take a loot at how the validation error behaves, but tfsummary files haven't been updated for more than a month.
2. Is training for a tiny number of steps helpful? Actual rating doesn't change linearly during the course of training, and it will certainly depend on the rating manifold in the weights space, which is not necessarily linear or very smooth. I am not sure if any of the networks trained for 8k steps has promoted, maybe a couple of them?

**Comment by Marcin1960** [comment]
@isty2e "I am not sure why a stricter gating is necessarily helpful. I don't think a small number of false positives (2.5%) is that harmful, and we also have larger probability of missing actual progress."

Unless I missed something, the "small number of false positives" might amount to even majority of promotions. Example, if there is NO real progress, you WILL  get the net  promotion just for this 2.5% or 10% - once in 40 or in 10 tried nets.

**Comment by isty2e** [comment]
@Marcin1960 Sure, but is it harmful? One can run matches between networks which are generations away from time to time, and AFAIK there has been no regression in strength so far.

**Comment by jkiliani** [comment]
Successful training step counts at the current schedule cannot be used to make any deductions about what step counts would work with the proposed adaptive window. What's currently happening (and what I presume is slowing down progress quite a bit) is:

After a successful promotion, we continue training, with the best net as starting point, but using a mix of games that were not yet trained on and those that were already trained on as a dataset. A very large part of the training effort, especially soon after a promotion, goes into training on games that were already trained on, which somewhat increases the correlation of positions that were used to train those network weights. This may not hurt in cases where not that many positions from those games were previously trained on, but if it's many positions, it will slightly increase the value head overfitting. When we do find a new net, I think that's usually at the point when the benefit from training on new games is much larger than the questionable and possibly negative use of training on more positions from older games.

**Comment by Marcin1960** [comment]
"When we do find a new net, I think that's usually at the point when the benefit from training on new games is much larger than the questionable and possibly negative use of training on more positions from older games."

Looks to me very sensible. One should be clever about managing of the game pool, especially that new games must come much slower as the network is much bigger. So this waste you are pointing to, needs to be reduced.

**Comment by isty2e** [comment]
Indeed networks experience the same (or highly correlated) data point. But let us do a quick math: 128k steps with batch size of 512 (which might be smaller for 192x15) and assuming the average game length is 128 moves, with a training window of 512k games. Then, one training cycle with 128k steps approximately corresponds to training for an epoch (or 0.5 epoch for a batch size of 256). However, it is very usual to train a model for several epochs, so the argument alone doesn't support that we should exclude already used games.

Of course there are uncertainties which are not considered here. The same training data point stays for many cycles, and there is correlation between moves in a single game, and also between moves from different games. On the other hand, the training games are constantly replaced, and not all promoted networks are from 128k steps.

What I'd like to say is that we don't know what is happening unless it is supported by data, such as tfsummary files. We do not know whether there is overfitting with current training configurations or not, and it is in fact rather surprising for me that many cares about overfitting but less about underfitting, which also should be monitored carefully.

**Comment by isty2e** [comment]
I do agree that overfitting is more likely to happen for the value head, but I do not quite understand "that would definitely happen by training only on very few positions per game" part. Do we have any data to discuss on?

**Comment by jkiliani** [comment]
OK, that statement is from the perspective that we have a limited pool of self-play data. The hypothetical ideal dataset is a pool of self-play games hundreds of millions large, so we could afford to only train on one position per game, with completely uncorrelated policy and value heads.

In practise we'll never have that luxury, so we have to train on multiple positions per game, to make the most out of the training data. "Most" just shouldn't be such much the the value head is overfit.

**Comment by isty2e** [comment]
I am still not sure if that supports your argument. I assume it might be that the correlation between values is larger than correlation between policies? At any rate, I don't think we have the data to determine whether there is overfitting or underfitting here. Overfitting can be detected by validation, and underfitting by reducing the loss function further and comparing, but we don't have any data to figure out what is happening now.

**Comment by jkiliani** [comment]
I did provide what data I could access for this argument. More data will of course be found in the training logs, but I can't access those for Leela Zero, only for Leela chess where they are accessible online at http://training.lczero.org/#scalars. What I observed there was that policy accuracy tends to eventually plateau but almost never oscillate at least on small learning rates, while test MSE tends to go up again eventually after many training steps. Not sure what kind of data you expect from me there, I analysed the data accessible to me and made a (imho) logical case. If you disagree with the arguments, please explain.

**Comment by jkiliani** [comment]
Fully agree that it would be great if we could browse through Tensorboard logs without having a working Tensorflow installation available, but I have to do with what I have.

In any case, if this discussion ends up dead as a result of 

> For the people who have a lot of ideas and like to argue: convincing, actionable data (or even better, code that can be tested for effectiveness) will make my opinion flip-flop like the best/worst politician, whereas arguing with words only is likely to be as fun and effective as slamming your head against a wall repeatedly.

Oh well, I tried.

**Comment by isty2e** [comment]
My position always has been that we can't tell much unless we have precisely the data what we need and have tested enough. Hopefully @gcp might update the tensorflow summary file...

**Comment by jkiliani** [comment]
And I agree with you in principle. Even if this only resulted in @gcp making Tensorboard as accessible as @Error323 for LCZero, that would already be great, because then more people could analyse what's going on exactly with the network training process of Leela Zero.

**Comment by hydrogenpi** [comment]
the op is crazy. 60℅ gating he is out of his freaking mind. we never get  60% anymore. what a joke

**Comment by Ishinoshita** [comment]
'@jkiliani - "There must be a certain number of positions per game that you can safely train on during one training run without overfitting the value head"
Not sure if this is relevant to the present discussion, but:

It strikes me ever since AGZ paper that the tree search result is used as a target only for policy head, value head been clamped to the +1/-1 game outcome target. That leads to sampling a single position from each game,  to avoid overfitting of value head. A value head target mixing, e.g. in a 50%/50% manner, game outcome and tree search evaluation at root, would yield a different target for each single position, opening the door to sampling multiple positions per game. At least once the value head has converged a bit. But whether this would improve/degrade the strength, I have no idea.

Also, I note that RL concept of importance sampling /prioritized replay is not considered in AGZ, LZ and LCZ, to focus the training on situations where the networks fails badly. As long as the error is high for both heads, there should be no risk of overfitting when using multiple time a position that the networks fail to evaluate correctly, no ? So that such positions could be oversampled, to speed up learning. ...Still a 'zero' approach, from go/ chess point of view, just the use of a general RL concept.

**Comment by hydrogenpi** [comment]
if anything gating needs to be lowered to 53℅ at 1000 games mark. any talks about raising gating isnt aligned with reality

**Comment by jkiliani** [comment]
@Ishinoshita I also thought of value head target mixing, but @gcp said that while it sounds logical, it's been tried and didn't work. I think he must have experimented with network training sufficiently developing Leela that he's right in that.

@hydrogenpi I get the impression you didn't get the concept I described here, only reacted on instinct to the 60% proposal. I already gave an explanation why I think it's a good idea to @herazul's prior objection.

By the way, I do think that this proposal would work well with two more modifications: Using t=1 for the whole game, and reducing visits to a lower value to speed up game generation, e.g. 1000 (+1 for the file compression ratio).

**Comment by hydrogenpi** [comment]
I read what you wrote, but the fact is at this level the glory days of 55℅ are over, we will never hit 60% naturally again , aka sans net2net upticks, and look at the jump to 15.192, we already flatining again and need to go to 20.256 just to get higher elo now.

**Comment by jkiliani** [comment]
There's only one reason we never hit 50% naturally again, and that's because we create and test a lot of nets. Before the strength of a network can reach 60% over the old one, a somewhat weaker one will have been promoted long ago, and become the new (stronger) reference.

This proposal is about both creating and selecting the networks we do test in a more efficient way. And by the way I think 192x15 is nowhere near to stalling, it's just a bit slow going because of the reduced game rate.

**Comment by Marcin1960** [comment]
@isty2e "My position always has been that we can't tell much unless we have precisely the data what we need and have tested enough."

Silence is gold :)

**Comment by Marcin1960** [comment]
@jkiliani " I think 192x15 is nowhere near to stalling, it's just a bit slow going because of the reduced game rate."

This is what I have been thinking. It is enough to look at the graph. LZ is going up, just the movement on the X axis got slower, so 3x more patience is required.

**Comment by hydrogenpi** [comment]
then we need more clients. how to get more people. boinc?

**Comment by hydrogenpi** [comment]
gold is gold. silence is not
evil prevails when good men fail to act

**Comment by hydrogenpi** [comment]
im not just looking at time, if on average it takes more than 20 nets to promote one, then thats sign of possible trouble, and this early on into a new network arch is not good imho

**Comment by Friday9i** [comment]
"im not just looking at time, if on average it takes more than 20 nets to promote one, then thats sign of possible trouble, and this early on into a new network arch is not good imho": possible, I've got that fear too.
But on the bright side, we still have very few high-quality games with the new 15x192 networks (less than 60K games as of now): hence, I've got the impression that too many networks were computed and tested with almost similar inputs, which leaves very little room to improve. In the coming days, with 200K and more games from 15x192 networks, things should improve much more hopefully ; -)

**Comment by Ishinoshita** [comment]
@jkiliani Thank you very much for your answer ! I trust facts and @gcp's experience.You calmed down one of my obsession ;-). Before I buy my own GPU and experiment by myself, I will now have to learn to live with my second obsession ;-), namely policy target enhancement (play self-play games with N playouts, sample positions and evaluate them through extra MCTS search with k*N playouts, e.g.k=10). Doesn't fit current training pipeline, I know. Expected benefit (offered with no proof): faster learning, particularly on deep tactical reading, at marginal additional computation cost/ game production rate drop (~ -5% for k=10, assuming 200 moves avg. game length).

**Comment by gcp** [comment]
@jkiliani if I can reword/modify what you are describing into something much simpler:

Do a dump of only very recent training data (I'd say, last 2 networks), do a very small training run (say 16k steps max?), and test networks after 2k steps, 4k steps, 8k steps, 16k steps.

The idea would be that a little bit of training on the newest information might be the most productive.

Practically, I think it interacts badly with SWA. A lot of the recent promotions have been very high step networks, and that's because they simply have better SWA averaging. So I think for this to work one has to solve the problem what to set the learning rate at.

**Comment by jkiliani** [comment]
It really depends on exactly how SWA is practised... I think the basic idea of an exponential step count testing schedule is sound, I'm just curious about the specifics.

**Comment by jkiliani** [comment]
> SWA is averaging every dump at 8k. My code matches what is in the repo. You could dump more frequently, but I'm not sure that's going to get you the same result as multiple averages that are many steps apart.

That would depends on how many training steps are required at quasi-stationary SGD (i.e. weights are already circling the optima, no longer searching for them) to reach maximum difference in consecutive weight dumps. The mean deviation between weights at a given step count difference should be an asymptotic function, with the asymptote only determined by learning rate (and indirectly, batch size). I have no idea what the decay factor would look like though, do you need only a couple hundred steps to circle the optima once on average, or several thousand? Has anyone looked into this yet?
